lib/fastlane/plugin/wpmreleasetoolkit/helper/strings_file_validation_helper.rb

```diff
@@ -0,0 +1,129 @@
+            state.buffer.string = ''
+            :after_quoted_key_before_eq
+          end,
+          /./ => lambda do |state, c|
+            state.buffer.write(c)
+            :in_quoted_key
+          end
+        },
+        after_quoted_key_before_eq: {
+          /\s/ => :after_quoted_key_before_eq,
+          '=' => :after_quoted_key_and_eq
+        },
+        after_quoted_key_and_eq: {
+          /\s/ => :after_quoted_key_and_eq,
+          '"' => :in_quoted_value
+        },
+        in_quoted_value: {
+          '"' => :after_quoted_value,
+          /./m => :in_quoted_value
+        },
+        after_quoted_value: {
+          /\s/ => :after_quoted_value,
+          ';' => :root
+        }
+      }.freeze
+
+      # Returns all the keys in the given `.strings` file, and the line(s) at
+      # which they are defined.
+      #
+      # @param [String] file The path to the file to inspect.
+      # @return [<Hash>] A `Hash` with keys the key values read from the
+      #         `.strings` file. The value of each entry is an array with the
+      #         line numbers where the key occurs.
+      def self.get_keys_from_strings(file:)
+        keys_with_lines = {}
+
+        state = State.new(context: :root, buffer: StringIO.new, in_escaped_ctx: false, found_key: nil)
+
+        File.readlines(file).each_with_index do |line, line_no|
+          line.chars.each_with_index do |c, col_no|
+            # Handle escaped characters at a global level. This is more
+            # straightforward than having a `TRANSITIONS` table that account
+            # for it.
+            if state.in_escaped_ctx || c == '\\'
+              # Just because we check for escaped characters at the global
+              # level, it doesn't mean we allow them in every context.
+              raise "Found escaped character outside of allowed contexts (current context: #{state.context})" unless %i[in_quoted_key in_quoted_value in_block_comment].include?(state.context)
+
+              state.buffer.write(c) if state.context == :in_quoted_key
+              state.in_escaped_ctx = !state.in_escaped_ctx
+              next
+            end
+
+            # Look at the transitions table for the current context, and find
+            # the first transition matching the current character
+            (_, next_context) = TRANSITIONS[state.context].find { |regex, _| c.match?(regex) } || [nil, nil]
+            raise "Invalid character `#{c}` found on line #{line_no + 1}, col #{col_no + 1}" if next_context.nil?
+
+            state.context = next_context.is_a?(Proc) ? next_context.call(state, c) : next_context
+            next unless state.found_key
+
+            # If we just exited the :in_quoted_key context and thus have found
+            # a new key, process it
+            key = state.found_key.dup
+            state.found_key = nil
+
+            keys_with_lines[key] = keys_with_lines[key].nil? ? [line_no + 1] : keys_with_lines[key].push(line_no + 1)
```

Comment:
Now that I think about it, I think there's a way in Ruby to provide [a default value when none is associated with a key](https://ruby-doc.org/core-3.1.1/Hash.html#method-c-new), something like:

```ruby
 # Use this on line 70 instead of the keys_with_lines = {}
keys_with_lines = Hash.new([])
…
# then here, that should simplify this code to just this, given that `keys_with:
keys_with_lines[key] += [line_no + 1]
```

_PS: Note that if we go that route, we would need to use `+= [x]` instead of `.push(x)` because `hash[key]` only **returns** the default value if the `hash` does not have one for `key`, but would not **set** that default value back in the `hash` (so `hash[key].push(x)` would just have `hash[key]` return a `[]` empty array instance, push a value `x` into it… then discard the result, without ever storing it back into `hash`, while `+=` does store it back._

Reply:
Brilliant! I'll try that.